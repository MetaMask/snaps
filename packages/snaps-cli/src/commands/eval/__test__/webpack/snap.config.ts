import type { SnapConfig } from '../../../../config';

const config: SnapConfig = {
  bundler: 'webpack',
  input: 'src/index.ts',
};

export default config;
