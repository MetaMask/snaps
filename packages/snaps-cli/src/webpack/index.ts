export * from './compiler';
export * from './config';
export * from './plugins';
