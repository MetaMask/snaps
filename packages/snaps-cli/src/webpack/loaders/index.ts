export * from './function';
export { default as browserify } from './browserify';
export { default as wasm } from './wasm';
