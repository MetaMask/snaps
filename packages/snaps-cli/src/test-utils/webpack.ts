import { logError } from '@metamask/snaps-utils';
import { DEFAULT_SNAP_BUNDLE } from '@metamask/snaps-utils/test-utils';
import type { IFs } from 'memfs';
import { createFsFromVolume, Volume } from 'memfs';
import { promisify } from 'util';
import type { Configuration } from 'webpack';
import webpack from 'webpack';

export type CompileOptions = {
  code?: string;
  fileSystem?: IFs;
  config?: Configuration;
  watchMode?: boolean;
};

/**
 * Compile a snap bundle using webpack.
 *
 * @param options - Options to configure the compilation.
 * @param options.code - The snap bundle code.
 * @param options.fileSystem - The file system to use. Defaults to an in-memory
 * file system.
 * @param options.config - The webpack configuration to use. Defaults to a
 * minimal configuration that compiles the provided code.
 * @param options.watchMode - Whether to compile in watch mode. Defaults to
 * false.
 * @returns The compiled code.
 */
export async function getCompiler({
  code = DEFAULT_SNAP_BUNDLE,
  fileSystem = createFsFromVolume(new Volume()),
  config,
  watchMode = false,
}: CompileOptions = {}) {
  await fileSystem.promises.writeFile('/input.js', code);

  const compiler = webpack({
    entry: '/input.js',
    mode: 'none',
    output: {
      path: '/',
      filename: 'output.js',
    },
    ...config,
  });

  compiler.watchMode = watchMode;
  compiler.inputFileSystem = fileSystem;
  compiler.outputFileSystem = fileSystem;

  return compiler;
}

/**
 * Compile a snap bundle using webpack.
 *
 * @param options - Options to configure the compilation.
 * @param options.code - The snap bundle code.
 * @param options.fileSystem - The file system to use. Defaults to an in-memory
 * file system.
 * @param options.config - The webpack configuration to use. Defaults to a
 * minimal configuration that compiles the provided code.
 * @returns The compiled code.
 */
export async function compile(
  options: CompileOptions,
): Promise<{ code: string }> {
  const compiler = await getCompiler(options);

  return new Promise((resolve, reject) => {
    compiler.run(async (error, stats) => {
      if (error || !stats) {
        reject(new Error('Failed to compile.'));
        return;
      }

      if (stats.hasErrors()) {
        logError(stats.toString('errors-only'));
        reject(new Error('Failed to compile.'));
        return;
      }

      const output = (await promisify(compiler.outputFileSystem.readFile)(
        '/output.js',
      )) as string;

      resolve({
        code: output,
      });
    });
  });
}

/**
 * Fakes the fallback modules path to ensure that snapshots have the same path on different systems.
 *
 * @param config - The webpack configuration.
 * @returns The webpack configuration with the fallback paths normalized.
 */
export function fakePolyfillPaths(config: Configuration): Configuration {
  if (!config.resolve?.fallback) {
    return config;
  }

  const {
    resolve: { fallback, ...resolveRest },
    ...rest
  } = config;

  const normalizedFallbacks = Object.keys(fallback).reduce((acc, index) => {
    if (typeof (fallback as Record<string, string>)[index] === 'string') {
      return { [index]: '/foo/bar/index.js', ...acc };
    }

    return { [index]: (fallback as Record<string, string>)[index], ...acc };
  }, {});

  return {
    ...rest,
    resolve: { ...resolveRest, fallback: normalizedFallbacks },
  };
}
