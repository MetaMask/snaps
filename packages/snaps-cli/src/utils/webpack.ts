import { logError, logInfo } from '@metamask/snaps-utils';
import SnapsWebpackPlugin from '@metamask/snaps-webpack-plugin';
import TerserPlugin from 'terser-webpack-plugin';
import { Compiler, Configuration, webpack } from 'webpack';

import { ProcessedWebpackConfig } from '../config';

export type WebpackOptions = {
  /**
   * Whether to watch for changes.
   */
  watch?: boolean;

  /**
   * Whether to evaluate the bundle. If this is set, it will override the
   * `evaluate` option in the config object.
   */
  evaluate?: boolean;
};

export type WatchPluginOptions = {
  /**
   * The extra files to watch.
   */
  files?: string[];
};

export class WatchPlugin {
  #options: WatchPluginOptions;

  constructor(options: WatchPluginOptions) {
    this.#options = options;
  }

  apply(compiler: Compiler) {
    compiler.hooks.afterCompile.tapPromise(
      'WatchPlugin',
      async ({ fileDependencies }) => {
        this.#options.files?.forEach(
          fileDependencies.add.bind(fileDependencies),
        );
      },
    );
  }
}

/**
 * Get the Webpack devtool configuration based on the given snap config.
 *
 * - If `sourceMap` is `inline`, return `inline-source-map`.
 * - If `sourceMap` is `true`, return `source-map`.
 * - Otherwise, return `false`.
 *
 * @param config - The processed snap Webpack config.
 * @param config.sourceMap - Whether to generate source maps.
 * @returns The Webpack devtool configuration.
 */
function getDevTool({
  sourceMap,
}: ProcessedWebpackConfig): Configuration['devtool'] {
  if (sourceMap === 'inline') {
    return 'inline-source-map';
  }

  if (sourceMap === true) {
    return 'source-map';
  }

  return false;
}

/**
 * Get the default Webpack configuration. This is the configuration that will
 * be used if the user doesn't provide a custom Webpack configuration. The
 * configuration is based on the snap config.
 *
 * The default configuration includes:
 *
 * - `SWC` to transpile TypeScript and JavaScript files.
 * - `TerserPlugin` to minify the bundle.
 * - `SnapsWebpackPlugin` to validate the bundle and update the manifest.
 *
 * It can be customized through the `customizeWebpackConfig` function in the
 * snap config, but in most cases, you shouldn't need to do that.
 *
 * @param config - The processed snap Webpack config.
 * @param options - The Webpack options.
 * @returns The default Webpack configuration.
 */
export function getDefaultConfiguration(
  config: ProcessedWebpackConfig,
  options: WebpackOptions = {},
): Configuration {
  return {
    /**
     * `web` is the default target for Webpack, so we don't need to set it, but
     * we do so here for clarity. This ensures that Webpack will build a bundle
     * that can be loaded in a browser.
     *
     * @see https://webpack.js.org/configuration/target/
     */
    target: 'web',

    /**
     * The mode is set to `production` by default, so that Webpack will minify
     * and optimize the bundle.
     *
     * @see https://webpack.js.org/configuration/mode/
     */
    mode: 'production',

    /**
     * The entry point is set to the `entry` value from the config object.
     *
     * @see https://webpack.js.org/configuration/entry-context/
     */
    entry: config.entry,

    /**
     * The devtool option controls how source maps are generated. We set it to
     * the `sourceMap` value from the config object.
     *
     * @see https://webpack.js.org/configuration/devtool/
     */
    devtool: getDevTool(config),

    /**
     * The stats option controls how much information is printed to the console
     * when Webpack is running. We set it to `none` so that we can control the
     * output ourselves.
     *
     * @see https://webpack.js.org/configuration/stats/
     */
    stats: 'none',

    /**
     * The output options.
     *
     * @see https://webpack.js.org/configuration/output/
     */
    output: {
      /**
       * This indicates whether Webpack should clear the output directory
       * before building. We set it to the `clean` value from the config
       * object.
       *
       * @see https://webpack.js.org/configuration/output/#outputclean
       */
      clean: config.output.clean,

      /**
       * The filename of the bundle. We set it to the `filename` value from
       * the config object.
       *
       * @see https://webpack.js.org/configuration/output/#outputfilename
       */
      filename: config.output.filename,

      /**
       * The path to the output directory. We set it to the `path` value from
       * the config object.
       *
       * @see https://webpack.js.org/configuration/output/#outputpath
       */
      path: config.output.path,

      /**
       * The public path of the bundle. We set it to `/` by default, so that
       * the bundle can be loaded from the root of the server.
       *
       * @see https://webpack.js.org/configuration/output/#outputpublicpath
       */
      publicPath: '/',

      /**
       * The library configuration. This tells Webpack how to export the
       * bundle.
       *
       * @see https://webpack.js.org/configuration/output/#outputlibrary
       */
      library: {
        /**
         * This tells Webpack to export the bundle as a CommonJS module. This
         * is necessary for MetaMask Snaps.
         *
         * @see https://webpack.js.org/configuration/output/#outputlibrarytarget
         */
        type: 'commonjs',
      },
    },

    /**
     * The module configuration. This is where we tell Webpack how to handle
     * different types of files.
     *
     * @see https://webpack.js.org/configuration/module/
     */
    module: {
      rules: [
        {
          /**
           * We use the `swc-loader` to transpile TypeScript and JavaScript
           * files. This is a Webpack loader that uses the `SWC` compiler,
           * which is a much faster alternative to Babel and TypeScript's own
           * compiler.
           *
           * @see https://swc.rs/docs/usage/swc-loader
           */
          test: /\.tsx?$/u,
          exclude: /node_modules/u,
          use: {
            loader: 'swc-loader',

            /**
             * The options for the `swc-loader`. These can be overridden in the
             * `.swcrc` file.
             *
             * @see https://swc.rs/docs/configuration/swcrc
             */
            options: {
              /**
               * This tells SWC to generate source maps. We set it to the
               * `sourceMap` value from the config object.
               *
               * This must be enabled if source maps are enabled in the config.
               */
              sourceMaps: Boolean(getDevTool(config)),

              jsc: {
                /**
                 * MetaMask targets ES2020, so we set the target to ES2020. If
                 * you need to support older browsers, you can set the target
                 * to something lower to include the necessary polyfills.
                 *
                 * @see https://swc.rs/docs/configuration/compilation#jsctarget
                 */
                target: 'es2020',

                parser: {
                  /**
                   * This tells the parser to parse TypeScript files. If you
                   * don't need to support TypeScript, you can set this to
                   * `ecmascript` instead, but there's no harm in leaving it
                   * as `typescript`.
                   *
                   * @see https://swc.rs/docs/configuration/compilation#jscparser
                   */
                  syntax: 'typescript',
                },
              },

              /**
               * The module configuration. This tells SWC how to output the
               * transpiled code.
               *
               * @see https://swc.rs/docs/configuration/modules
               */
              module: {
                /**
                 * This tells SWC to output CommonJS modules. MetaMask Snaps
                 * doesn't support ES modules yet, so this is necessary.
                 *
                 * @see https://swc.rs/docs/configuration/modules#commonjs
                 */
                type: 'commonjs',
              },
            },
          },
        },
      ],
    },

    /**
     * The resolve configuration. This tells Webpack how to resolve imports.
     * We set it to resolve `.js` and `.ts` files.
     *
     * @see https://webpack.js.org/configuration/resolve/
     */
    resolve: {
      /**
       * The extensions to resolve. We set it to resolve `.js` and `.ts`
       * files.
       */
      extensions: ['.js', '.ts'],
    },

    /**
     * The plugins to use.
     *
     * @see https://webpack.js.org/configuration/plugins/
     */
    plugins: [
      /**
       * The `SnapsWebpackPlugin` is a Webpack plugin that checks and updates
       * the manifest file, and evaluates the bundle in SES. While not strictly
       * required, it's highly recommended to use this plugin.
       */
      new SnapsWebpackPlugin({
        manifestPath: config.manifest.path,
        writeManifest: config.manifest.update,
        eval: options.evaluate ?? config.evaluate,
      }),

      /**
       * The `WatchPlugin` is a Webpack plugin that adds extra files to watch
       * for changes. This is useful for rebuilding the bundle when the
       * manifest file changes.
       */
      new WatchPlugin({
        files: [config.manifest.path],
      }),
    ],

    /**
     * The optimization configuration. This tells Webpack how to optimize the
     * bundle. Most of the time, you won't need to change this, as the default
     * options set by the `mode` option are sufficient.
     */
    optimization: {
      /**
       * The minimizer to use. We set it to use the `TerserPlugin`, which
       * minifies the bundle using SWC.
       */
      minimizer: [
        new TerserPlugin({
          minify: TerserPlugin.swcMinify,
        }),
      ],
    },

    /**
     * The infrastructure logging configuration. This tells Webpack how to
     * log messages.
     *
     * @see https://webpack.js.org/configuration/infrastructure-logging
     */
    infrastructureLogging: {
      /**
       * The level of logging to use. We set it to `none`, so that we can
       * control the output ourselves.
       */
      level: 'none',
    },
  };
}

/**
 * Get a Webpack compiler for the given config.
 *
 * @param config - The config object.
 * @param options - The Webpack options.
 * @returns The Webpack compiler.
 */
export function getCompiler(
  config: ProcessedWebpackConfig,
  options?: WebpackOptions,
) {
  const baseWebpackConfig = getDefaultConfiguration(config, options);
  const webpackConfig =
    config.customizeWebpackConfig?.(baseWebpackConfig) ?? baseWebpackConfig;

  const compiler = webpack(webpackConfig);
  compiler.hooks.afterDone.tap('AfterCompilePlugin', (stats) => {
    if (!stats) {
      return;
    }

    const { modules, time, errors } = stats.toJson();
    if (!modules || !time) {
      logError('Compilation status unknown. Please check your config.');
      process.exitCode = 1;
      return;
    }

    if (errors?.length) {
      logError(
        `Compiled ${modules?.length} files in ${time}ms with ${errors?.length} errors.`,
      );
      process.exitCode = 1;
      return;
    }

    logInfo(`Compiled ${modules?.length} files in ${time}ms.`);
  });

  return compiler;
}
