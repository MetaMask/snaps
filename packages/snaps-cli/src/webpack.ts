import SnapsWebpackPlugin from '@metamask/snaps-webpack-plugin';
import TerserPlugin from 'terser-webpack-plugin';
import { Configuration } from 'webpack';

import { ProcessedWebpackConfig } from './config';

/**
 * Get the Webpack devtool configuration based on the given snap config.
 *
 * - If `sourceMap` is `inline`, return `inline-source-map`.
 * - If `sourceMap` is `true`, return `source-map`.
 * - Otherwise, return `false`.
 *
 * @param config - The processed snap Webpack config.
 * @param config.sourceMap - Whether to generate source maps.
 * @returns The Webpack devtool configuration.
 */
function getDevTool({
  sourceMap,
}: ProcessedWebpackConfig): Configuration['devtool'] {
  if (sourceMap === 'inline') {
    return 'inline-source-map';
  }

  if (sourceMap === true) {
    return 'source-map';
  }

  return false;
}

/**
 * Get the default Webpack configuration. This is the configuration that will
 * be used if the user doesn't provide a custom Webpack configuration. The
 * configuration is based on the snap config.
 *
 * The default configuration includes:
 *
 * - `SWC` to transpile TypeScript and JavaScript files.
 * - `TerserPlugin` to minify the bundle.
 * - `SnapsWebpackPlugin` to validate the bundle and update the manifest.
 *
 * It can be customized through the `customizeWebpackConfig` function in the
 * snap config, but in most cases, you shouldn't need to do that.
 *
 * @param config - The processed snap Webpack config.
 * @returns The default Webpack configuration.
 */
export function getDefaultConfiguration(
  config: ProcessedWebpackConfig,
): Configuration {
  return {
    /**
     * `web` is the default target for Webpack, so we don't need to set it, but
     * we do so here for clarity. This ensures that Webpack will build a bundle
     * that can be loaded in a browser.
     *
     * @see https://webpack.js.org/configuration/target/
     */
    target: 'web',

    /**
     * The mode is set to `production` by default, so that Webpack will minify
     * and optimize the bundle.
     *
     * @see https://webpack.js.org/configuration/mode/
     */
    mode: 'production',

    /**
     * The entry point is set to the `entry` value from the config object.
     *
     * @see https://webpack.js.org/configuration/entry-context/
     */
    entry: config.entry,

    /**
     * The devtool option controls how source maps are generated. We set it to
     * the `sourceMap` value from the config object.
     *
     * @see https://webpack.js.org/configuration/devtool/
     */
    devtool: getDevTool(config),

    /**
     * The output options.
     *
     * @see https://webpack.js.org/configuration/output/
     */
    output: {
      /**
       * This indicates whether Webpack should clear the output directory
       * before building. We set it to the `clean` value from the config
       * object.
       *
       * @see https://webpack.js.org/configuration/output/#outputclean
       */
      clean: config.output.clean,

      /**
       * The filename of the bundle. We set it to the `filename` value from
       * the config object.
       *
       * @see https://webpack.js.org/configuration/output/#outputfilename
       */
      filename: config.output.filename,

      /**
       * The path to the output directory. We set it to the `path` value from
       * the config object.
       *
       * @see https://webpack.js.org/configuration/output/#outputpath
       */
      path: config.output.path,

      /**
       * The library configuration. This tells Webpack how to export the
       * bundle.
       *
       * @see https://webpack.js.org/configuration/output/#outputlibrary
       */
      library: {
        /**
         * This tells Webpack to export the bundle as a CommonJS module. This
         * is necessary for MetaMask Snaps.
         *
         * @see https://webpack.js.org/configuration/output/#outputlibrarytarget
         */
        type: 'commonjs',
      },
    },

    /**
     * The module configuration. This is where we tell Webpack how to handle
     * different types of files.
     *
     * @see https://webpack.js.org/configuration/module/
     */
    module: {
      rules: [
        {
          /**
           * We use the `swc-loader` to transpile TypeScript and JavaScript
           * files. This is a Webpack loader that uses the `SWC` compiler,
           * which is a much faster alternative to Babel and TypeScript's own
           * compiler.
           *
           * @see https://swc.rs/docs/usage/swc-loader
           */
          test: /\.tsx?$/u,
          exclude: /node_modules/u,
          use: {
            loader: 'swc-loader',

            /**
             * The options for the `swc-loader`. These can be overridden in the
             * `.swcrc` file.
             *
             * @see https://swc.rs/docs/configuration/swcrc
             */
            options: {
              parseMap: true,
              sourceMaps: true,
              inputSourceMap: true,
              /**
               * This tells SWC to generate source maps. We set it to the
               * `sourceMap` value from the config object.
               *
               * This must be enabled if source maps are enabled in the config.
               */
              // sourceMaps: Boolean(getDevTool(config)),

              jsc: {
                /**
                 * MetaMask targets ES2020, so we set the target to ES2020. If
                 * you need to support older browsers, you can set the target
                 * to something lower to include the necessary polyfills.
                 *
                 * @see https://swc.rs/docs/configuration/compilation#jsctarget
                 */
                target: 'es2020',

                parser: {
                  /**
                   * This tells the parser to parse TypeScript files. If you
                   * don't need to support TypeScript, you can set this to
                   * `ecmascript` instead, but there's no harm in leaving it
                   * as `typescript`.
                   *
                   * @see https://swc.rs/docs/configuration/compilation#jscparser
                   */
                  syntax: 'typescript',
                },
              },

              /**
               * The module configuration. This tells SWC how to output the
               * transpiled code.
               *
               * @see https://swc.rs/docs/configuration/modules
               */
              module: {
                /**
                 * This tells SWC to output CommonJS modules. MetaMask Snaps
                 * doesn't support ES modules yet, so this is necessary.
                 *
                 * @see https://swc.rs/docs/configuration/modules#commonjs
                 */
                type: 'commonjs',
              },
            },
          },
        },
      ],
    },

    /**
     * The resolve configuration. This tells Webpack how to resolve imports.
     * We set it to resolve `.js` and `.ts` files.
     *
     * @see https://webpack.js.org/configuration/resolve/
     */
    resolve: {
      /**
       * The extensions to resolve. We set it to resolve `.js` and `.ts`
       * files.
       */
      extensions: ['.js', '.ts'],
    },

    /**
     * The plugins to use.
     *
     * @see https://webpack.js.org/configuration/plugins/
     */
    plugins: [
      new SnapsWebpackPlugin({
        manifestPath: config.manifest.path,
        writeManifest: config.manifest.update,
      }),
    ],

    /**
     * The optimization configuration. This tells Webpack how to optimize the
     * bundle. Most of the time, you won't need to change this, as the default
     * options set by the `mode` option are sufficient.
     */
    optimization: {
      /**
       * The minimizer to use. We set it to use the `TerserPlugin`, which
       * minifies the bundle using SWC.
       */
      minimizer: [
        new TerserPlugin({
          minify: TerserPlugin.swcMinify,
        }),
      ],
    },
  };
}
