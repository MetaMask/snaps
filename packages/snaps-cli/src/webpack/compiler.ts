import { logError, logInfo } from '@metamask/snaps-utils';
import { resolve } from 'path';
import { webpack } from 'webpack';
import WebpackDevServer from 'webpack-dev-server';

import { ProcessedWebpackConfig } from '../config';
import { getDefaultConfiguration, WebpackOptions } from './config';

/**
 * Get a Webpack compiler for the given config.
 *
 * @param config - The config object.
 * @param options - The Webpack options.
 * @returns The Webpack compiler.
 */
export function getCompiler(
  config: ProcessedWebpackConfig,
  options?: WebpackOptions,
) {
  const baseWebpackConfig = getDefaultConfiguration(config, options);
  const webpackConfig =
    config.customizeWebpackConfig?.(baseWebpackConfig) ?? baseWebpackConfig;

  const compiler = webpack(webpackConfig);
  compiler.hooks.afterDone.tap('AfterCompilePlugin', (stats) => {
    if (!stats) {
      return;
    }

    const { modules, time, errors } = stats.toJson();
    if (!modules || !time) {
      logError('Compilation status unknown. Please check your config.');
      process.exitCode = 1;
      return;
    }

    if (errors?.length) {
      logError(
        `Compiled ${modules?.length} files in ${time}ms with ${errors?.length} errors.`,
      );
      process.exitCode = 1;
      return;
    }

    logInfo(`Compiled ${modules?.length} files in ${time}ms.`);
  });

  return compiler;
}

/**
 * Get a Webpack dev server for the given config.
 *
 * @param config - The config object.
 * @param options - The Webpack options.
 * @returns The Webpack dev server.
 */
export function getServer(
  config: ProcessedWebpackConfig,
  options: WebpackOptions,
) {
  const compiler = getCompiler(config, options);

  return new WebpackDevServer(
    {
      port: config.server.port,
      static: {
        directory: resolve(process.cwd(), config.server.root),
        watch: options.watch,
        serveIndex: true,
      },

      devMiddleware: {
        writeToDisk: true,
      },

      magicHtml: false,
      hot: false,
      liveReload: false,

      client: {
        logging: 'error',
        overlay: false,
        progress: false,
        reconnect: false,
      },

      onListening: () => {
        logInfo(`Server listening on: http://localhost:${config.server.port}`);
      },
    },
    compiler,
  );
}
