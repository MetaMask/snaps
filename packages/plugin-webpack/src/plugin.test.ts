// Allow Jest snapshots because the test outputs are illegible.
/* eslint-disable jest/no-restricted-matchers */

import webpack, { Stats } from 'webpack';
import { createFsFromVolume, Volume } from 'memfs';
import SnapsWebpackPlugin from './plugin';

describe('SnapsWebpackPlugin', () => {
  it('processes files using Webpack', async () => {
    const fileSystem = createFsFromVolume(new Volume());
    const { promises: fs } = fileSystem;

    const bundler = webpack({
      mode: 'none',
      entry: {
        foo: '/foo.js',
      },
      output: {
        path: '/lib',
        filename: '[name].js',
      },
      plugins: [new SnapsWebpackPlugin()],
    });

    bundler.inputFileSystem = fileSystem;
    bundler.outputFileSystem = fileSystem;

    await fs.mkdir('/lib', { recursive: true });
    await fs.writeFile('/foo.js', `const foo = 'bar';`);

    await new Promise<Stats>((resolve, reject) =>
      bundler.run((error, stats) => {
        if (error || !stats) {
          return reject(error);
        }

        return resolve(stats);
      }),
    );

    const result = await fs.readFile('/lib/foo.js', 'utf-8');

    expect(result).toMatchSnapshot();
  });

  it('applies a transform', async () => {
    const fileSystem = createFsFromVolume(new Volume());
    const { promises: fs } = fileSystem;

    const bundler = webpack({
      mode: 'none',
      entry: {
        foo: '/foo.js',
      },
      output: {
        path: '/lib',
        filename: '[name].js',
      },
      plugins: [new SnapsWebpackPlugin()],
    });

    bundler.inputFileSystem = fileSystem;
    bundler.outputFileSystem = fileSystem;

    await fs.mkdir('/lib', { recursive: true });
    await fs.writeFile(
      '/foo.js',
      `
        // foo bar
        /* baz qux */
        const foo = 'bar';
     `,
    );

    await new Promise<Stats>((resolve, reject) =>
      bundler.run((error, stats) => {
        if (error || !stats) {
          return reject(error);
        }

        return resolve(stats);
      }),
    );

    const result = await fs.readFile('/lib/foo.js', 'utf-8');

    expect(result).toMatchSnapshot();
    expect(result).not.toContain(`// foo bar`);
    expect(result).not.toContain(`/* baz qux */`);
  });

  it('forwards the options', async () => {
    const fileSystem = createFsFromVolume(new Volume());
    const { promises: fs } = fileSystem;

    const bundler = webpack({
      mode: 'none',
      entry: {
        foo: '/foo.js',
      },
      output: {
        path: '/lib',
        filename: '[name].js',
      },
      plugins: [new SnapsWebpackPlugin({ stripComments: false })],
    });

    bundler.inputFileSystem = fileSystem;
    bundler.outputFileSystem = fileSystem;

    await fs.mkdir('/lib', { recursive: true });
    await fs.writeFile(
      '/foo.js',
      `
        // foo bar
        /* baz qux */
        const foo = 'bar';
     `,
    );

    await new Promise<Stats>((resolve, reject) =>
      bundler.run((error, stats) => {
        if (error || !stats) {
          return reject(error);
        }

        return resolve(stats);
      }),
    );

    const result = await fs.readFile('/lib/foo.js', 'utf-8');

    expect(result).toMatchSnapshot();
    expect(result).toContain(`// foo bar`);
    expect(result).toContain(`/* baz qux */`);
  });

  it('runs on the entire bundle', async () => {
    const fileSystem = createFsFromVolume(new Volume());
    const { promises: fs } = fileSystem;

    const bundler = webpack({
      mode: 'none',
      entry: {
        foo: '/foo.js',
      },
      output: {
        path: '/lib',
        filename: '[name].js',
      },
      plugins: [new SnapsWebpackPlugin()],
    });

    bundler.inputFileSystem = fileSystem;
    bundler.outputFileSystem = fileSystem;

    await fs.mkdir('/lib', { recursive: true });

    await fs.writeFile(
      '/foo.js',
      `
        import { bar } from './bar';

        // Sets foo to bar
        const foo = bar;
     `,
    );

    await fs.writeFile(
      '/bar.js',
      `
        // Returns baz
        export const bar = 'baz';
     `,
    );

    await new Promise<Stats>((resolve, reject) =>
      bundler.run((error, stats) => {
        if (error || !stats) {
          return reject(error);
        }

        return resolve(stats);
      }),
    );

    const result = await fs.readFile('/lib/foo.js', 'utf-8');

    expect(result).toMatchSnapshot();
    expect(result).not.toContain(`// Sets foo to bar`);
    expect(result).not.toContain(`// Returns baz`);
  });
});
