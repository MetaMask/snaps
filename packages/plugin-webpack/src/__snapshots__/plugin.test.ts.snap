// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnapsWebpackPlugin applies a transform 1`] = `
" (() => { 
var __webpack_exports__ = {};

        
                const foo = 'bar';
     
 })()
;"
`;

exports[`SnapsWebpackPlugin forwards the options 1`] = `
"/******/ (() => { // webpackBootstrap
var __webpack_exports__ = {};

        // foo bar
        /* baz qux */
        const foo = 'bar';
     
/******/ })()
;"
`;

exports[`SnapsWebpackPlugin processes files using Webpack 1`] = `
" (() => { 
var __webpack_exports__ = {};
const foo = 'bar';
 })()
;"
`;

exports[`SnapsWebpackPlugin runs on the entire bundle 1`] = `
" (() => { 
 	\\"use strict\\";
 	var __webpack_modules__ = ([
,
 ((__unused_webpack_module, __webpack_exports__, __webpack_require__) => {

__webpack_require__.r(__webpack_exports__);
 __webpack_require__.d(__webpack_exports__, {
   \\"bar\\": () => ( bar)
 });

        
        const bar = 'baz';
     

 })
 	]);
 	
 	var __webpack_module_cache__ = {};
 	
 	
 	function __webpack_require__(moduleId) {
 		
 		var cachedModule = __webpack_module_cache__[moduleId];
 		if (cachedModule !== undefined) {
 			return cachedModule.exports;
 		}
 		
 		var module = __webpack_module_cache__[moduleId] = {
 			
 			
 			exports: {}
 		};
 	
 		
 		__webpack_modules__[moduleId](module, module.exports, __webpack_require__);
 	
 		
 		return module.exports;
 	}
 	
 	 	(() => {
 		
 		__webpack_require__.d = (exports, definition) => {
 			for(var key in definition) {
 				if(__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
 					Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });
 				}
 			}
 		};
 	})();
 	
 	 	(() => {
 		__webpack_require__.o = (obj, prop) => (Object.prototype.hasOwnProperty.call(obj, prop))
 	})();
 	
 	 	(() => {
 		
 		__webpack_require__.r = (exports) => {
 			if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
 				Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
 			}
 			Object.defineProperty(exports, '__esModule', { value: true });
 		};
 	})();
 	
var __webpack_exports__ = {};

(() => {
__webpack_require__.r(__webpack_exports__);
 var _bar__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(1);

        

        
        const foo = _bar__WEBPACK_IMPORTED_MODULE_0__.bar;
     
})();

 })()
;"
`;
